src/Core/hco_calc_mod.F90

```diff
@@ -2679,25 +2687,73 @@ SUBROUTINE GetDilFact( HcoState,   EmisL1, EmisL1Unit, EmisL2,             &
+       ! that the same amount of mass is emitted into each layer.
+       IF ( HcoState%Options%VertWeight ) THEN 
 
-       ! Adjust dh if we are in top level
-       IF ( L == UppLL ) THEN
-          dh = h2 - SUM(HcoState%Grid%BXHEIGHT_M%Val(I,J,1:(UppLL-1)))
-       ENDIF
+          ! Height of grid box of interest (in m)
+          dh = HcoState%Grid%BXHEIGHT_M%Val(I,J,L)
 
-       ! compute dilution factor: the new flux should emit the same mass per
-       ! volume, i.e. flux_total/column_total = flux_level/column_level
-       ! --> flux_level = fluxtotal * column_level / column_total.
-       IF ( h2 > h1 ) THEN
-          DilFact = dh / ( h2 - h1 )
+          ! Adjust dh if we are in lowest level
+          IF ( L == LowLL ) THEN
+             dh = SUM(HcoState%Grid%BXHEIGHT_M%Val(I,J,1:LowLL)) - h1
+          ENDIF
+
+          ! Adjust dh if we are in top level
+          IF ( L == UppLL ) THEN
+             dh = h2 - SUM(HcoState%Grid%BXHEIGHT_M%Val(I,J,1:(UppLL-1)))
+          ENDIF
+
+          ! compute dilution factor: the new flux should emit the same mass per
+          ! volume, i.e. flux_total/column_total = flux_level/column_level
+          ! --> flux_level = fluxtotal * column_level / column_total.
+          IF ( h2 > h1 ) THEN
+             DilFact = dh / ( h2 - h1 )
+          ELSE
+             MSG = 'GetDilFact h2 not greater than h1'
+             CALL HCO_ERROR ( HcoState%Config%Err, MSG, RC, THISLOC=LOC )
+             RETURN
+          ENDIF
+
+       ! If VertWeight option is turned off, emit the same flux in each layer.
+       ! Since model layers have different depths, this will result in differnt
+       ! total emissions per layer.
        ELSE
+#ifdef MODEL_GEOS
+          ! Get fractional layer indeces for lower and upper level. This makes 
+          ! sure that only fractions of the lower and upper level are being
+          ! considered, so that double-counting is avoided if a model layer serves
+          ! both as the top layer and the bottom layer (e.g., wildfire emissions
+          ! emitted from bottom to the top of PBL, and from the top of PBL to 5000m).
+          LowLLR = REAL(LowLL,hp) - 1.0_hp
+          UppLLR = REAL(UppLL,hp)
+          dh1 = 0.0 
+          DO L1 = 1, HcoState%NZ
+             dh2 = SUM(HcoState%Grid%BXHEIGHT_M%Val(I,J,1:L1))
+             IF ( h1 >= dh1 .AND. h1 < dh2 ) THEN
+                LowLLR = REAL(L1,hp) - ( (dh2-h1)/(dh2-dh1) )
+             ENDIF      
+             IF ( h2 > dh1 .AND. h2 <= dh2 ) THEN
+                UppLLR = REAL(L1,hp) - ( (dh2-h2)/(dh2-dh1) )
+             ENDIF      
+             ! top layer is bottom layer in next loop
+             dh1 = dh2 
+          ENDDO
+
+          ! Dilution factor using fractional levels
+          IF ( UppLLR <= LowLLR ) THEN
+             DilFact = 1.0_hp / REAL(UppLL-LowLL+1,hp)
+          ELSE
+             DilFact = 1.0_hp / (UppLLR-LowLLR)
+          ENDIF
+
+#else
```

Comment:
Adding the error message here is not correct. To maintain the old behavior (outside of MODEL_GEOS), it should be:
`DilFact = 1.0_hp / REAL(UppLL-LowLL+1,hp)`

However, I'd recommend to add this update to the standard code, rather than using a MODEL_GEOS bracket. The new code is more accurate for cases where you specify fixed emission depths (e.g., if you want to emit everything into the bottom 5000 meters).

Reply:
Got it. I removed the message that crept in. I also removed the `MODEL_GEOS` ifdef so that the new handling for when `HcoState%Grid%BXHEIGHT_M%Val` is associated and `HcoState%Options%VertWeight `is false becomes standard.